geo_tag/geo_tag.py

```diff
@@ -0,0 +1,221 @@
+                    raise ValueError("Invalid coordinates in bounding_box.")
+
+                # Return Central Point or Random Point in the polygon each time.
+                if mode == 0:
+                    return ((swLog+neLog) / 2, (swLat+neLat) / 2)
+                elif mode == 1:
+                    poly = Polygon([(swLog, swLat), (neLog, swLat), (neLog, neLat), (swLog, neLat)])
+                    while True:
+                        random_point = Point([random.uniform(swLog, neLog), random.uniform(swLat, neLat)])
+                        if random_point.within(poly):
+                            return random_point
+                else:
+                    raise ValueError("Invalid mode selection in bounding_box.")
+
+    # private: help function. city_state_name: eg. (Irvine, CA). flag: source of infer.
+    def __extract_geo_tag_from_city_and_state(self, tweet_json: dict, coord: tuple, city_state_name: str, flag: str):
+        if city_state_name is None:
+            raise ValueError("Can not find city_name in extract function.")
+
+        if "," not in city_state_name:
+            raise ValueError("Invalid city_name to parse in extract function.")
+
+        state_of_city = city_state_name.split(",")[1]
+        target_state = state_of_city.strip()
+        full_state_name = self.abbrev_us_state.get(target_state)
+
+        if full_state_name is None:
+            raise KeyError("Can not find state_name in extract function.")
+
+        # search it in the cache(hashmap) to get the county's name
+        target_key = city_state_name.split(",")[0] + ', ' + full_state_name
+
+        if coord is not None:
+            # coord is empty now
+            location = self.geo_locator.geocode(target_key)
+            coord = (location.longitude, location.latitude)
+
+        geo_content = self.cache.get(target_key)
+        if geo_content is not None:
+            geo_tag = dict()
+            geo_tag["stateID"] = geo_content[2]
+            geo_tag["stateName"] = full_state_name
+            geo_tag["countyID"] = geo_content[1]
+            geo_tag["countyName"] = geo_content[3]
+            geo_tag["cityID"] = geo_content[0]
+            geo_tag["cityName"] = city_state_name.split(",")[0]
+            geo_tag["coordinate"] = coord
+            geo_tag["source"] = flag
+
+            tweet_json['geo_tag'] = geo_tag
+            print(geo_tag)
+            
+        else:
+            raise ValueError("Not find county.")
+
+    def __infer_geo_from_place(self, tweet_json: dict, coord: tuple):
+        place = tweet_json.get('place')
+        if place is not None:
+            # extract the city and state abbrev
+            city_state_name = place.get('full_name')
+            if city_state_name is None:
+                try:
+                    raise KeyError()
+                except KeyError:
+                    print("Not find full_name key in place.")
+                return
```

Comment:
what is this for? raise the exception to outer scope, you are still handling in this if block...

Reply:
Sorry, don't notice this one.